src/storage-client/src/storage_collections.rs

```diff
@@ -519,10 +534,86 @@ where
         since: Option<&Antichain<T>>,
         relation_desc: RelationDesc,
         persist_client: &PersistClient,
-    ) -> (
-        WriteHandle<SourceData, (), T, Diff>,
-        SinceHandle<SourceData, (), T, Diff, PersistEpoch>,
-    ) {
+    ) -> (WriteHandle<SourceData, (), T, Diff>, SinceHandleWrapper<T>) {
+        let since_handle = if self.read_only.load(Ordering::SeqCst) {
+            let read_handle = self
+                .open_leased_handle(id, shard, relation_desc.clone(), since, persist_client)
+                .await;
+            SinceHandleWrapper::Leased(read_handle)
+        } else {
+            let since_handle = self
+                .open_critical_handle(id, shard, since, persist_client)
+                .await;
+
+            SinceHandleWrapper::Critical(since_handle)
+        };
+
+        let mut write_handle = self
+            .open_write_handle(id, shard, relation_desc, persist_client)
+            .await;
+
+        // N.B.
+        // Fetch the most recent upper for the write handle. Otherwise, this may
+        // be behind the since of the since handle. Its vital this happens AFTER
```

Comment:
Can you explain what goes wrong if the upper of the write handle is stale?

Reply:
All I can offer is https://github.com/MaterializeInc/materialize/pull/22740, which links to a GH issue and an internal slack discussion.

The tl;dr is:
- There is an assertion that fires when the upper is not beyond the since: https://github.com/MaterializeInc/materialize/issues/22725
  - I don't know if this is a good assertion to have, but 🤷‍♂️ 
- Creating a `WriteHandle` is not linearized with other things _and_ does not fetch a recent upper. Hence we have this extra `fetch_recent_upper()`.

Reply:
The first version of this PR had it inside `open_write_handle`, but then the comment about making sure that `fetch_recent_upper` didn't make sense anymore, because that specifically says we want to fetch the upper _after_ opening the since handle. So I moved it back to this outer method, where we can actually make sure it happens the way the comment describes.

Reply:
> There is an assertion that fires when the upper is not beyond the since

Looks like that assert was removed in this monster PR: https://github.com/MaterializeInc/materialize/pull/26556, fwiw.

Reply:
Yeah, I mentioned above that I don't know that this is/was actually a good assertion. But, that's for some other PR to figure out, I'm afraid 🙈